compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -339,14 +339,20 @@ impl<'a> Parser<'a> {
             HelpIdentifierStartsWithNumber { num_span: invalid }
         });
 
-        let err = ExpectedIdentifier {
-            span: bad_token.span,
-            token: bad_token,
-            suggest_raw,
-            suggest_remove_comma,
-            help_cannot_start_number,
+        let mut err = if let Some(lo) = self.maybe_ternary_lo && self.token == token::Colon {
+            self.bump();
+            self.parse_expr()?;
```

Comment:
Can you [snapshot](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_parse/parser/struct.Parser.html#method.create_snapshot_for_diagnostic) the parser before the bump and recover `Err` from `parse_expr` by canceling the error, restoring the snapshot, and falling into the branch below?

I'd like for this ternary error to not result in some even more cursed knockdown errors.

Reply:
Would this be necessary in `maybe_recover_from_ternary_operator` as well?

Reply:
Yeah... hm... ideally those two places wouldn't be doing almost the same exact logic as well...

Reply:
Yeah it's unfortunate, but `expected_ident_found` seems to be necessary as otherwise it'll sometimes complain about finding `:` instead of an identifier. I'm not entirely sure why the other check in `expected_one_of_not_found` doesn't cover it :/

Parsing the second expression should be extracted though